File: solutions/ruby/anagram/anagram.rb
class Anagram
  def initialize(target)
    @base = target.downcase
    @base_sorted = @base.chars.sort
  end

  def is_anagram?(candidate)
    return false unless candidate.length == @base.length
    word = candidate.downcase
    return false if word == @base
    return word.chars.sort == @base_sorted
  end

  def match(candidates)
    candidates.select { |word| is_anagram?(word) }
  end
end

File: solutions/csharp/anagram/Anagram.cs
using System;
using System.Linq;

public class Anagram
{
    private string subject;
    private char[] baseLetters;

    public Anagram(string baseWord)
    {
        subject = baseWord.ToLower();
        baseLetters = sorted(subject);
    }

    public string[] FindAnagrams(string[] potentialMatches) =>
        // returns all candidates that are anagrams of, but not equal to, 'subject'
        potentialMatches.Where(word =>
        {
            if (word.Length != subject.Length) return false;

            var candidate = word.ToLower();
            if (candidate == subject) return false;

            return sorted(candidate).SequenceEqual(baseLetters);

        }).ToArray();

    private char[] sorted(string input)
    {
        var letters = input.ToCharArray();
        Array.Sort(letters);
        return letters;
    }
}